### PR TITLE
AM-3522 Group mapper on idp level

### PR DESCRIPTION
**Pull request by lgw-gravitee**

### CODE REVIEW

File: gravitee-am-identityprovider/gravitee-am-identityprovider-api/src/main/java/io/gravitee/am/identityprovider/api/DefaultIdentityProviderGroupMapper.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    public static final Logger LOGGER = LoggerFactory.getLogger(DefaultIdentityProviderGroupMapper.class);
+
+
+    @Override
+    public Map<String, String[]> getGroups() {
+        return getProperties();
+    }
+
+    @Override
+    public void setGroups(Map<String, String[]> roles) {
```
Comment (leleueri) [inline gravitee-am-identityprovider/gravitee-am-identityprovider-api/src/main/java/io/gravitee/am/identityprovider/api/DefaultIdentityProviderGroupMapper.java:33]:
```suggestion
    public void setGroups(Map<String, String[]> groups) {
```

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/user/BaseUserEnhancer.java [hunk 2]
```diff
@@ -37,9 +37,16 @@ public abstract class BaseUserEnhancer implements UserEnhancer {
     protected final Logger LOGGER = LoggerFactory.getLogger(getClass());
     protected abstract Flowable<Group> getGroupsByMemberId(String memberId);
     protected abstract Flowable<Role> getRolesByIds(List<String> roleIds);
+    protected abstract Flowable<Group> getGroupsByIds(List<String> groupIds);
+
+    private Flowable<Group> collectGroups(User user) {
+        Flowable<Group> groupsByMemberId = getGroupsByMemberId(user.getId());
+        Flowable<Group> groupsByIds = getGroupsByIds(user.getDynamicGroups());
+        return groupsByMemberId.mergeWith(groupsByIds);
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/user/BaseUserEnhancer.java:45]: question: should we ensure somehow uniquess in the final Flowable as merged flows may have duplicates ?

**Review by leleueri — COMMENTED** [review]
Linter is failing on the UI

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/user/BaseUserEnhancer.java (hunk 2 above)
Comment (lgw-gravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/user/BaseUserEnhancer.java:45]: good point. I convert List to Set in the end.

**Review by leleueri — APPROVED** [review]
I tested a simple case and It is working + code looks ok.
I created a task to add groupMapper on each EE IDP (AM-3874 to not block this one)

**Pull request merged**